platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsAccessFromKotlinDslIntegrationTest.groovy

```diff
@@ -89,12 +90,18 @@ class IsolatedProjectsAccessFromKotlinDslIntegrationTest extends AbstractIsolate
         """
 
         when:
-        isolatedProjectsFails(":a:help", ":b:help")
+        // TODO:isolated expected behavior for incremental configuration
+//        isolatedProjectsFails(":a:help")
+        isolatedProjectsRun(":a:help")
 
         then:
-        fixture.assertStateStoredAndDiscarded {
+        // TODO:isolated expected behavior for incremental configuration
+//        fixture.assertStateStoredAndDiscarded {
+//            projectsConfigured(":", ":a")
+//            problem("Build file 'a/build.gradle.kts': Project ':a' cannot access 'Project.buildDir' functionality on another project ':b'")
```

Comment:
🤔 Does it mean the `gradle.beforeProject` lambda is not executed at all for the projects that happen to be configured at the time of the invocation? If so, this seems like a very unintuitive behavior that is also not noted in the javadoc.

Reply:
Yes, when the call to `gradle.beforeProject` happens after the project has been evaluated.

I wonder if that's not already the case Today when you have the call in a project that's been configured after another (e.g. in project `:b` instead of `:a`) 🤔

Reply:
The implementation of the `gradle.beforeProject` callbacks relies on listeners, which are probably not "stateful", so they are okay with late listeners not receiving earlier events. This is an unintuitive contract.

Ideally, we should forbid calling this method after projects have loaded, same as we forbid calling `gradle.lifecycle.beforeProject` from the beginning.

Reply:
With IP, shouldn't `gradle.beforeProject` itself be a violation, period?

Reply:
It probably should. I was also suggesting updating the vintage behavior.

Reply:
I'm not so sure about changing the behavior of `gradle.beforeProject` in vintage. That would probably translate into it always being an error to call it from a project script since, by definition, no script would be able to install a callback before the root project is evaluated, for example.